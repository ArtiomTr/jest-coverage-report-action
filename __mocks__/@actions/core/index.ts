const input: Record<string, unknown> = {};

export const getInput = (name: string) => input[name];

export const mockInput = (newInput: Record<string, unknown>) => {
    Object.assign(input, newInput);
};

export const clearInputMock = () => {
    Object.keys(input).map((key) => {
        delete input[key];
    });
};

export const error = jest.fn();
export const info = jest.fn();
export const setFailed = jest.fn();
