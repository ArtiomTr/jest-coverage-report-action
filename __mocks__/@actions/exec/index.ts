import { vi } from 'vitest';

export const exec = vi.fn();
