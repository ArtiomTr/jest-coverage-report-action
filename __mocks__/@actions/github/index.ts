export const context: Record<string, unknown> = {};

export const mockContext = (newContext: Record<string, unknown>) => {
    Object.assign(context, newContext);
};

export const clearContextMock = () => {
    Object.keys(context).map((key) => {
        delete context[key];
    });
};

export const getOctokit = jest.fn();
