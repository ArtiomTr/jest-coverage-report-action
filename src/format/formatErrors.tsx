import { toMarkdown } from 'mdast-util-to-markdown';

import { Errors } from '../components/Errors.js';

export const formatErrors = (errors: Array<Error>) => {
    return toMarkdown(
        <root>
            <Errors errors={errors} />
        </root>,
        {
            fences: true,
        }
    );
};
