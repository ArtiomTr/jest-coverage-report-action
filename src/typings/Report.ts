export enum FailReason {
    TESTS_FAILED = 'testsFailed',
    INVALID_COVERAGE_FORMAT = 'invalidFormat',
    UNDER_THRESHOLD = 'underThreshold',
    UNKNOWN_ERROR = 'unknownError',
    REPORT_NOT_FOUND = 'reportNotFound',
    READING_COVERAGE_FILE_FAILED = 'readingCoverageFileFailed',
    FAILED_GETTING_COVERAGE = 'failedGettingCoverage',
}

export type TestRunReport = {
    title: string;
    summary: string;
    failures?: string;
};

export type SummaryReport = {
    text: string;
    runReport: TestRunReport;
};
