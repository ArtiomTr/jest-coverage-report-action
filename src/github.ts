import { context, getOctokit } from '@actions/github';

import {
    PullRequestInfo,
    PullRequestInfoQuery,
    PullRequestInfoQueryVariables,
} from './generated/graphql';

export type Octokit = ReturnType<typeof getOctokit>;

function graphql<TQuery, TQueryVariables extends Record<string, unknown>>(
    octokit: Octokit,
    source: string,
    variables: TQueryVariables
) {
    return octokit.graphql<TQuery>(source, {
        ...variables,
    });
}

export const getPullRequestInfo = async (
    octokit: Octokit,
    prNumber?: number
) => {
    if (context?.payload?.pull_request) {
        return context?.payload?.pull_request;
    }

    if (!prNumber) {
        return;
    }

    const output = await graphql<
        PullRequestInfoQuery,
        PullRequestInfoQueryVariables
    >(octokit, PullRequestInfo, {
        repo: context.repo.repo,
        owner: context.repo.owner,
        prNumber,
    });

    output.repository?.pullRequest?.baseRef?.repository.nameWithOwner;
};
