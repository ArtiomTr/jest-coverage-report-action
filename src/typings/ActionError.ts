import { FailReason } from './Report';
import { i18n } from '../utils/i18n';

export class ActionError<T> extends Error {
    public constructor(reason: FailReason, details?: T) {
        super(
            i18n(
                `errors.${reason}`,
                (details as unknown) as Record<string, unknown>
            )
        );
    }

    public toString(): string {
        return this.message;
    }
}
