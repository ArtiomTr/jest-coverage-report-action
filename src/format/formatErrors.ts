import { ActionError } from '../typings/ActionError';
import { getConsoleLink } from '../utils/getConsoleLink';
import { i18n } from '../utils/i18n';

const getNumberWidth = (index: number) => Math.floor(Math.log10(index));

export const formatErrors = (errors: Array<Error>) => {
    if (errors.length === 0) {
        return '';
    }

    if (errors.length === 1) {
        const error = errors[0];

        if (error instanceof ActionError) {
            return i18n(':x: {{ error }}', { error: error.toString() });
        }

        return i18n(':x: {{ unexpectedError }} \n```\n{{ error }}\n```', {
            error: error.toString(),
            unexpectedError: i18n('errors.unexpectedError', {
                consoleLink: getConsoleLink(),
            }),
        });
    }

    return (
        i18n('errors.multiple') +
        i18n('\n```\n{{ errors }}\n```\n', {
            errors: errors
                .map(
                    (error, index) =>
                        ` ${String(1 + index).padEnd(
                            1 + getNumberWidth(errors.length),
                            ' '
                        )} | ${error.toString()}`
                )
                .join('\n'),
        })
    );
};
