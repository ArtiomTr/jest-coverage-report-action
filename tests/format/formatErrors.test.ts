import * as all from '@actions/github';

import { formatErrors } from '../../src/format/formatErrors';
import { ActionError } from '../../src/typings/ActionError';
import { FailReason } from '../../src/typings/Report';

const { mockContext, clearContextMock } = all as any;

describe('formatErrors', () => {
    it('should return empty string, if no errors specified', () => {
        expect(formatErrors([])).toBe('');
    });

    it('should format 1 error', () => {
        expect(formatErrors([new ActionError(FailReason.TESTS_FAILED)])).toBe(
            '❌ The test suite failed. Please, check the console output for more details.'
        );

        mockContext({
            payload: {},
            repo: {
                owner: 'bot',
                repo: 'test-repo',
            },
            runId: 10,
        });

        expect(formatErrors([new Error('This is unexpected error')])).toBe(
            '❌ An unexpected error occurred. For more details, [check console](https://github.com/bot/test-repo/actions/runs/10) ' +
                '\n```\n' +
                'Error: This is unexpected error' +
                '\n```'
        );

        clearContextMock();
    });

    it('should format multiple errors', () => {
        expect(
            formatErrors([
                new ActionError(FailReason.TESTS_FAILED),
                new Error('hello'),
                new ActionError(FailReason.TESTS_FAILED),
                new EvalError('aaa'),
            ])
        ).toMatchSnapshot();

        expect(
            formatErrors(new Array(40).fill(0).map(() => new Error('error')))
        ).toMatchSnapshot();

        expect(
            formatErrors(new Array(100).fill(0).map(() => new Error('error')))
        ).toMatchSnapshot();
    });
});
