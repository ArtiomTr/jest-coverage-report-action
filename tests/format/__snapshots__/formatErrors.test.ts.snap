// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatErrors > should format multiple errors 1`] = `
"Multiple errors occurred
\`\`\`
 1 | The test suite failed. Please, check the console output for more details.
 2 | Error: hello
 3 | The test suite failed. Please, check the console output for more details.
 4 | EvalError: aaa
\`\`\`
"
`;

exports[`formatErrors > should format multiple errors 2`] = `
"Multiple errors occurred
\`\`\`
 1  | Error: error
 2  | Error: error
 3  | Error: error
 4  | Error: error
 5  | Error: error
 6  | Error: error
 7  | Error: error
 8  | Error: error
 9  | Error: error
 10 | Error: error
 11 | Error: error
 12 | Error: error
 13 | Error: error
 14 | Error: error
 15 | Error: error
 16 | Error: error
 17 | Error: error
 18 | Error: error
 19 | Error: error
 20 | Error: error
 21 | Error: error
 22 | Error: error
 23 | Error: error
 24 | Error: error
 25 | Error: error
 26 | Error: error
 27 | Error: error
 28 | Error: error
 29 | Error: error
 30 | Error: error
 31 | Error: error
 32 | Error: error
 33 | Error: error
 34 | Error: error
 35 | Error: error
 36 | Error: error
 37 | Error: error
 38 | Error: error
 39 | Error: error
 40 | Error: error
\`\`\`
"
`;

exports[`formatErrors > should format multiple errors 3`] = `
"Multiple errors occurred
\`\`\`
 1   | Error: error
 2   | Error: error
 3   | Error: error
 4   | Error: error
 5   | Error: error
 6   | Error: error
 7   | Error: error
 8   | Error: error
 9   | Error: error
 10  | Error: error
 11  | Error: error
 12  | Error: error
 13  | Error: error
 14  | Error: error
 15  | Error: error
 16  | Error: error
 17  | Error: error
 18  | Error: error
 19  | Error: error
 20  | Error: error
 21  | Error: error
 22  | Error: error
 23  | Error: error
 24  | Error: error
 25  | Error: error
 26  | Error: error
 27  | Error: error
 28  | Error: error
 29  | Error: error
 30  | Error: error
 31  | Error: error
 32  | Error: error
 33  | Error: error
 34  | Error: error
 35  | Error: error
 36  | Error: error
 37  | Error: error
 38  | Error: error
 39  | Error: error
 40  | Error: error
 41  | Error: error
 42  | Error: error
 43  | Error: error
 44  | Error: error
 45  | Error: error
 46  | Error: error
 47  | Error: error
 48  | Error: error
 49  | Error: error
 50  | Error: error
 51  | Error: error
 52  | Error: error
 53  | Error: error
 54  | Error: error
 55  | Error: error
 56  | Error: error
 57  | Error: error
 58  | Error: error
 59  | Error: error
 60  | Error: error
 61  | Error: error
 62  | Error: error
 63  | Error: error
 64  | Error: error
 65  | Error: error
 66  | Error: error
 67  | Error: error
 68  | Error: error
 69  | Error: error
 70  | Error: error
 71  | Error: error
 72  | Error: error
 73  | Error: error
 74  | Error: error
 75  | Error: error
 76  | Error: error
 77  | Error: error
 78  | Error: error
 79  | Error: error
 80  | Error: error
 81  | Error: error
 82  | Error: error
 83  | Error: error
 84  | Error: error
 85  | Error: error
 86  | Error: error
 87  | Error: error
 88  | Error: error
 89  | Error: error
 90  | Error: error
 91  | Error: error
 92  | Error: error
 93  | Error: error
 94  | Error: error
 95  | Error: error
 96  | Error: error
 97  | Error: error
 98  | Error: error
 99  | Error: error
 100 | Error: error
\`\`\`
"
`;
